std/experimental/algorithm.d

```diff
@@ -0,0 +1,49 @@
+// Written in the D programming language.
+
+/**
+This module implements experimental additions/modifications to $(MREF std, _algorithm).
+
+Use this module to test out new functionality for $(REF wrap, std, _algorithm)
+
+Source:    $(PHOBOSSRC std/experimental/_algorithm.d)
+
+Copyright: Copyright the respective authors, 2008-
+License:   $(HTTP boost.org/LICENSE_1_0.txt, Boost License 1.0).
+Authors:   Andrea Fontana
+           Razvan Nitu
+ */
+module std.experimental.algorithm;
+
+import std.range : isInputRange, ElementType;
+
+/**
+Lazily produces a range with no duplicated elements from the input
+range $(D r). The order of elements is the original order.
+
+Params:
+    Range r - the input range to eliminate duplicates from
+
+Returns:
+    A range containing all the unique elements
+*/
+auto distinct(Range)(Range r)
+if (isInputRange!Range)
+{
+    import std.algorithm.iteration : filter;
+    bool[ElementType!Range] justSeen;
```

Comment:
When I suggested moving this to `stdx.algorithm`, I was suggesting that we can use this to experiment with APIs that allow the user to control how the allocation happens.
There are a couple of ways to do this. Some quick ideas (there are more!):
- Do the deallocation automatically, e.g. like RefCounted and provide an optional Allocator
- Let the user do the deallocation (e.g. by returning the range and the allocated array as tuple)

Please note that it's very important that this function can be used in `@nogc`.